src/shared/tpm2-util.c

```diff
@@ -924,6 +924,18 @@ static int tpm2_make_policy_session(
 
         log_debug("Starting authentication session.");
 
+        /* So apparently some TPM implementations don't implement trial mode correctly. To avoid issues let's
+         * avoid it when it is easy to. At the moment we only really need trial mode for the signed PCR
+         * policies (since only then we need to shove PCR values into the policy that don't match current
+         * state anyway), hence if we have none of those we don't need to bother. Hence, let's patch in
```

Comment:
Does this imply that with normal PCR binding, one can't bind to a different PCR value than what is already there?

For example if PCR 15 is used to mix volume(s) encryption keys, then how to bind to PCR 15? After boot it will already be in unclean state. While for root volume it has to be pristine to be able to unlock.

Reply:
it doesn't make too much sense to bind your root volume to PCR 15, because it's supposed to carry a hash of the root volume key... So you'd read the PCR15 measurement before that, but then it's typically all zeroes, so why bother.

Reply:
Reason to do is that attacker may bott with same kernel and initrd but replace encrypted root volume with a rogue encrypted volume with same id. Unlock it with *password* (because TPM will apparently fail) and boot with PCRs in a state that attacker can unlock the real encrypted volume. But if PCR 15 (or another one) is already messed up, then that will not be possible.

i.e. make sure no volume was unlocked previously.

btw in Fedora 38 crypttab I have options
```
discard,tpm2-device=auto,tpm2-measure-pcr=yes,tpm2-pcrs=0+1+4+5+7+15
```

But then `tpm2_pcrread` still returns:
```
    15: 0x0000000000000000000000000000000000000000000000000000000000000000
```

So I'm not sure this works at all. But I don't have `systemd-stub` installed. Could that be the reason?

Is there a howto or at least all steps that are needed to get binding working for PCRs that changed after volume was unlocked?

It doesn't have to be PCR 15. For example initrd may have some logic to mess up other PCRs after unlock. Or another example is PCR 11 where we have:
>       systemd-pcrphase.service, systemd-pcrphase-sysinit.service, and
>       systemd-pcrphase-initrd.service are system services that measure
>       specific strings into TPM2 PCR 11 during boot at various milestones of
>       the boot process.

Reply:
please direct questions like this to systemd-devel btw. Adding discussion to a long merged PR really sucks.

The PCR 15 measurements are only done if the system is booted via sd-stub. See the man page for details. And read ukify man page how to generate them, as mentioned elsewhere.